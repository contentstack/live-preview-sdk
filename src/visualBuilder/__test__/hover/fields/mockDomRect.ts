export const mockDomRect = {
    singleHorizontal: () => ({
        bottom: 88.3984375,
        height: 54.3984375,
        left: 34,
        right: 862,
        top: 34,
        width: 828,
        x: 34,
        y: 34,
    }),
    singleLeft: () => ({
        x: 51,
        y: 51,
        width: 27.7734375,
        height: 20.3984375,
        top: 51,
        bottom: 71.3984375,
        left: 51,
        right: 78.7734375,
    }),
    singleRight: () => ({
        x: 110.7734375,
        y: 51,
        width: 53.59375,
        height: 20.3984375,
        top: 51,
        bottom: 71.3984375,
        left: 110.7734375,
        right: 164.3671875,
    }),
    singleVertical: () => ({
        bottom: 195.1953125,
        height: 90.796875,
        left: 34,
        right: 862,
        top: 104.3984375,
        width: 828,
        x: 34,
        y: 104.3984375,
    }),
    singleTop: () => ({
        x: 51,
        y: 121.3984375,
        width: 794,
        height: 20.3984375,
        top: 121.3984375,
        bottom: 141.796875,
        left: 51,
        right: 845,
    }),
    singleBottom: () => ({
        x: 51,
        y: 157.796875,
        width: 794,
        height: 20.3984375,
        top: 157.796875,
        bottom: 178.1953125,
        left: 51,
        right: 845,
    }),
    multipleLeft: () => ({
        x: 51,
        y: 228.1953125,
        width: 87.59375,
        height: 90.796875,
        top: 228.1953125,
        bottom: 318.9921875,
        left: 51,
        right: 138.59375,
    }),
    multipleRight: () => ({
        x: 170.59375,
        y: 228.1953125,
        width: 87.59375,
        height: 90.796875,
        top: 228.1953125,
        bottom: 318.9921875,
        left: 170.59375,
        right: 258.1875,
    }),
    multipleTop: () => ({
        x: 51,
        y: 368.9921875,
        width: 794,
        height: 90.796875,
        top: 368.9921875,
        bottom: 459.7890625,
        left: 51,
        right: 845,
    }),
    multipleBottom: () => ({
        x: 51,
        y: 475.7890625,
        width: 794,
        height: 90.796875,
        top: 475.7890625,
        bottom: 566.5859375,
        left: 51,
        right: 845,
    }),
    multipleChildTop: () => ({
        x: 68,
        y: 385.9921875,
        width: 760,
        height: 20.3984375,
        top: 385.9921875,
        bottom: 406.390625,
        left: 68,
        right: 828,
    }),
    multipleChildBottom: () => ({
        x: 68,
        y: 422.390625,
        width: 760,
        height: 20.3984375,
        top: 422.390625,
        bottom: 442.7890625,
        left: 68,
        right: 828,
    }),
};
