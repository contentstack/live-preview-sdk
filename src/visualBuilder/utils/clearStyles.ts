const originalStyles = new WeakMap<
    HTMLElement,
    {
        visibility?: string;
        transition?: string;
        animation?: string;
    }
>();

export function clearVisibelityStyles(element: HTMLElement) {
    const originalStyleValues = {
        visibility: element.style.visibility,
        transition: element.style.transition,
        animation: element.style.animation,
    };

    originalStyles.set(element, originalStyleValues);

    element.style.visibility = "hidden";
    element.style.transition = "none";
    element.style.animation = "none";
}

export function restoreVisibelityStyles(element: HTMLElement) {
    const storedStyles = originalStyles.get(element);

    if (storedStyles) {
        element.style.visibility = storedStyles.visibility || "";
        element.style.transition = storedStyles.transition || "";
        element.style.animation = storedStyles.animation || "";

        originalStyles.delete(element);
    }
}
