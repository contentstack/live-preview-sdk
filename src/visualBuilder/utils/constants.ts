import { FieldDataType } from "./types/index.types";

export const numericInputRegex = /^-?\d*(\.\d*)?([eE][-+]?\d*)?$/;

export const VISUAL_BUILDER_FIELD_TYPE_ATTRIBUTE_KEY = "data-cslp-field-type";

export const VISUAL_BUILDER_CHANNEL_ID = "visual-builder";

export const LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX = 2;

// Minimum distance from top edge to prevent toolbar from being hidden
export const TOP_EDGE_BUFFER = 42;
export const RIGHT_EDGE_BUFFER = 180;

export const TOOLBAR_EDGE_BUFFER = 8;

export const DATA_CSLP_ATTR_SELECTOR = "data-cslp";

/**
 * The field that can be directly modified using contenteditable=true.
 * This includes all text fields like title and numbers.
 */
export const ALLOWED_INLINE_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.SINGLELINE,
    FieldDataType.MULTILINE,
    FieldDataType.NUMBER,
];

export const ALLOWED_MODAL_EDITABLE_FIELD: FieldDataType[] = [
    FieldDataType.HTML_RTE,
    FieldDataType.MARKDOWN_RTE,
    FieldDataType.JSON_RTE,
    FieldDataType.CUSTOM_FIELD,
    FieldDataType.LINK,
    FieldDataType.ISODATE,
    FieldDataType.URL,
];

export const ALLOWED_REPLACE_FIELDS: FieldDataType[] = [
    FieldDataType.REFERENCE,
    FieldDataType.FILE,
];

export const DEFAULT_MULTIPLE_FIELDS: FieldDataType[] = [
    FieldDataType.GLOBAL_FIELD,
    FieldDataType.GROUP,
    FieldDataType.BLOCK,
];

export const mentionLimit = 20;
export const maxMessageLength = 500;
