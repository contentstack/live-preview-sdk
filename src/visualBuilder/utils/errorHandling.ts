
export function hasPostMessageError(obj: any): boolean {
    return obj?.error === true;
}