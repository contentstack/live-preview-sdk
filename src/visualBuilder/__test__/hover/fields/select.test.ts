import { getFieldSchemaMap } from "../../../../__test__/data/fieldSchemaMap";
import { waitForHoverOutline } from "../../../../__test__/utils";
import Config from "../../../../configManager/configManager";
import { VisualBuilder } from "../../../index";
import { FieldSchemaMap } from "../../../utils/fieldSchemaMap";
import { mockDomRect } from "./mockDomRect";
import { act, screen } from "@testing-library/preact";

vi.mock("../../../utils/visualBuilderPostMessage", async () => {
    const { getAllContentTypes } = await vi.importActual<
        typeof import("../../../../__test__/data/contentType")
    >("../../../../__test__/data/contentType");
    const contentTypes = getAllContentTypes();
    return {
        __esModule: true,
        default: {
            send: vi.fn().mockImplementation((eventName: string) => {
                if (eventName === "init")
                    return Promise.resolve({
                        contentTypes,
                    });
                return Promise.resolve();
            }),
        },
    };
});

global.ResizeObserver = vi.fn().mockImplementation(() => ({
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
}));

vi.mock("../../../../utils/index.ts", async () => {
    const actual = await vi.importActual("../../../../utils");
    return {
        __esModule: true,
        ...actual,
        isOpenInBuilder: vi.fn().mockReturnValue(true),
    };
});

describe("When an element is hovered in visual builder mode", () => {
    let mousemoveEvent: Event;

    beforeAll(() => {
        FieldSchemaMap.setFieldSchema(
            "all_fields",
            getFieldSchemaMap().all_fields
        );
        global.ResizeObserver = vi.fn().mockImplementation(() => ({
            observe: vi.fn(),
            unobserve: vi.fn(),
            disconnect: vi.fn(),
        }));

        global.MutationObserver = vi.fn().mockImplementation(() => ({
            observe: vi.fn(),
            disconnect: vi.fn(),
        }));
    });

    beforeEach(() => {
        Config.reset();
        Config.set("mode", 2);
        mousemoveEvent = new Event("mousemove", {
            bubbles: true,
            cancelable: true,
        });
        document.getElementsByTagName("html")[0].innerHTML = "";
    });

    afterEach(() => {
        vi.clearAllMocks();
        document.getElementsByTagName("html")[0].innerHTML = "";
    });

    afterAll(() => {
        Config.reset();
    });

    describe("select field", () => {
        let selectField: HTMLParagraphElement;
        let visualBuilder: VisualBuilder;

        beforeEach(() => {
            selectField = document.createElement("p");
            selectField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.select"
            );

            selectField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            document.body.appendChild(selectField);

            visualBuilder = new VisualBuilder();
        });

        afterEach(() => {
            visualBuilder.destroy();
        });

        test("should have outline and custom cursor", async () => {
            await act(async () => {
                selectField.dispatchEvent(mousemoveEvent);
            });
            await waitForHoverOutline();
            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute("style");

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute("data-icon", "select");
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });
    });

    describe("select field (multiple)", () => {
        let container: HTMLDivElement;
        let firstSelectField: HTMLParagraphElement;
        let secondSelectField: HTMLParagraphElement;
        let visualBuilder: VisualBuilder;

        beforeEach(() => {
            container = document.createElement("div");
            container.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.select_multiple_"
            );
            container.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleHorizontal());

            firstSelectField = document.createElement("p");
            firstSelectField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.select_multiple_.0"
            );

            firstSelectField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            secondSelectField = document.createElement("p");
            secondSelectField.setAttribute(
                "data-cslp",
                "all_fields.bltapikey.en-us.select_multiple_.1"
            );

            secondSelectField.getBoundingClientRect = vi
                .fn()
                .mockReturnValue(mockDomRect.singleLeft());

            container.appendChild(firstSelectField);
            container.appendChild(secondSelectField);
            document.body.appendChild(container);

            visualBuilder = new VisualBuilder();
        });

        afterEach(() => {
            visualBuilder.destroy();
        });

        test("should have outline and custom cursor", async () => {
            await act(async () => {
                container.dispatchEvent(mousemoveEvent);
            });
            await waitForHoverOutline();

            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute("style");

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute("data-icon", "select");
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });

        test("should have outline and custom cursor on individual instances", async () => {
            await act(async () => {
                firstSelectField.dispatchEvent(mousemoveEvent);
            });
            await waitForHoverOutline();

            const hoverOutline = document.querySelector(
                "[data-testid='visual-builder__hover-outline']"
            );
            expect(hoverOutline).toHaveAttribute("style");

            const customCursor = document.querySelector(
                `[data-testid="visual-builder__cursor"]`
            );

            expect(customCursor).toHaveAttribute("data-icon", "select");
            expect(customCursor?.classList.contains("visible")).toBeTruthy();
        });
    });
});
