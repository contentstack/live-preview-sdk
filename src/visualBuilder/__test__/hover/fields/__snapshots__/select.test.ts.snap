// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`When an element is hovered in visual builder mode > select field (multiple) > should have custom cursor 1`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="select"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper go4230334221"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 8.8125C2.89645 8.8125 2.8125 8.89645 2.8125 9V10.5C2.8125 10.8107 2.56066 11.0625 2.25 11.0625C1.93934 11.0625 1.6875 10.8107 1.6875 10.5V9C1.6875 8.27513 2.27513 7.6875 3 7.6875H4.5C4.81066 7.6875 5.0625 7.93934 5.0625 8.25C5.0625 8.56066 4.81066 8.8125 4.5 8.8125H3ZM6.9375 8.25C6.9375 7.93934 7.18934 7.6875 7.5 7.6875H10.5C10.8107 7.6875 11.0625 7.93934 11.0625 8.25C11.0625 8.56066 10.8107 8.8125 10.5 8.8125H7.5C7.18934 8.8125 6.9375 8.56066 6.9375 8.25ZM12.9375 8.25C12.9375 7.93934 13.1893 7.6875 13.5 7.6875H16.5C16.8107 7.6875 17.0625 7.93934 17.0625 8.25C17.0625 8.56066 16.8107 8.8125 16.5 8.8125H13.5C13.1893 8.8125 12.9375 8.56066 12.9375 8.25ZM18.9375 8.25C18.9375 7.93934 19.1893 7.6875 19.5 7.6875H21C21.7249 7.6875 22.3125 8.27513 22.3125 9V10.6875C22.3125 10.9982 22.0607 11.25 21.75 11.25C21.4393 11.25 21.1875 10.9982 21.1875 10.6875V9C21.1875 8.89645 21.1036 8.8125 21 8.8125H19.5C19.1893 8.8125 18.9375 8.56066 18.9375 8.25ZM2.25 12.9375C2.56066 12.9375 2.8125 13.1893 2.8125 13.5V15C2.8125 15.1036 2.89645 15.1875 3 15.1875H4.125C4.43566 15.1875 4.6875 15.4393 4.6875 15.75C4.6875 16.0607 4.43566 16.3125 4.125 16.3125H3C2.27513 16.3125 1.6875 15.7249 1.6875 15V13.5C1.6875 13.1893 1.93934 12.9375 2.25 12.9375ZM21.75 13.5C22.0607 13.5 22.3125 13.7518 22.3125 14.0625V15.75C22.3125 16.0607 22.0607 16.3125 21.75 16.3125C21.4393 16.3125 21.1875 16.0607 21.1875 15.75V14.0625C21.1875 13.7518 21.4393 13.5 21.75 13.5ZM5.8125 15.75C5.8125 15.4393 6.06434 15.1875 6.375 15.1875H8.625C8.93566 15.1875 9.1875 15.4393 9.1875 15.75C9.1875 16.0607 8.93566 16.3125 8.625 16.3125H6.375C6.06434 16.3125 5.8125 16.0607 5.8125 15.75ZM10.3125 15.75C10.3125 15.4393 10.5643 15.1875 10.875 15.1875H12C12.3107 15.1875 12.5625 15.4393 12.5625 15.75C12.5625 16.0607 12.3107 16.3125 12 16.3125H10.875C10.5643 16.3125 10.3125 16.0607 10.3125 15.75Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M13.4457 13.3706C13.3603 12.5628 14.2727 12.036 14.9296 12.5139L21.2889 17.1406C21.9906 17.6511 21.6818 18.7589 20.8172 18.8328L17.7404 19.0958L15.9742 21.6289C15.4779 22.3407 14.3642 22.0543 14.2729 21.1913L13.4457 13.3706ZM14.6089 13.6718L15.3388 20.5732L16.8678 18.3803C17.027 18.1519 17.2796 18.0061 17.557 17.9823L20.2207 17.7547L14.6089 13.6718Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > select field (multiple) > should have custom cursor 2`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="select"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper go4230334221"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 8.8125C2.89645 8.8125 2.8125 8.89645 2.8125 9V10.5C2.8125 10.8107 2.56066 11.0625 2.25 11.0625C1.93934 11.0625 1.6875 10.8107 1.6875 10.5V9C1.6875 8.27513 2.27513 7.6875 3 7.6875H4.5C4.81066 7.6875 5.0625 7.93934 5.0625 8.25C5.0625 8.56066 4.81066 8.8125 4.5 8.8125H3ZM6.9375 8.25C6.9375 7.93934 7.18934 7.6875 7.5 7.6875H10.5C10.8107 7.6875 11.0625 7.93934 11.0625 8.25C11.0625 8.56066 10.8107 8.8125 10.5 8.8125H7.5C7.18934 8.8125 6.9375 8.56066 6.9375 8.25ZM12.9375 8.25C12.9375 7.93934 13.1893 7.6875 13.5 7.6875H16.5C16.8107 7.6875 17.0625 7.93934 17.0625 8.25C17.0625 8.56066 16.8107 8.8125 16.5 8.8125H13.5C13.1893 8.8125 12.9375 8.56066 12.9375 8.25ZM18.9375 8.25C18.9375 7.93934 19.1893 7.6875 19.5 7.6875H21C21.7249 7.6875 22.3125 8.27513 22.3125 9V10.6875C22.3125 10.9982 22.0607 11.25 21.75 11.25C21.4393 11.25 21.1875 10.9982 21.1875 10.6875V9C21.1875 8.89645 21.1036 8.8125 21 8.8125H19.5C19.1893 8.8125 18.9375 8.56066 18.9375 8.25ZM2.25 12.9375C2.56066 12.9375 2.8125 13.1893 2.8125 13.5V15C2.8125 15.1036 2.89645 15.1875 3 15.1875H4.125C4.43566 15.1875 4.6875 15.4393 4.6875 15.75C4.6875 16.0607 4.43566 16.3125 4.125 16.3125H3C2.27513 16.3125 1.6875 15.7249 1.6875 15V13.5C1.6875 13.1893 1.93934 12.9375 2.25 12.9375ZM21.75 13.5C22.0607 13.5 22.3125 13.7518 22.3125 14.0625V15.75C22.3125 16.0607 22.0607 16.3125 21.75 16.3125C21.4393 16.3125 21.1875 16.0607 21.1875 15.75V14.0625C21.1875 13.7518 21.4393 13.5 21.75 13.5ZM5.8125 15.75C5.8125 15.4393 6.06434 15.1875 6.375 15.1875H8.625C8.93566 15.1875 9.1875 15.4393 9.1875 15.75C9.1875 16.0607 8.93566 16.3125 8.625 16.3125H6.375C6.06434 16.3125 5.8125 16.0607 5.8125 15.75ZM10.3125 15.75C10.3125 15.4393 10.5643 15.1875 10.875 15.1875H12C12.3107 15.1875 12.5625 15.4393 12.5625 15.75C12.5625 16.0607 12.3107 16.3125 12 16.3125H10.875C10.5643 16.3125 10.3125 16.0607 10.3125 15.75Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M13.4457 13.3706C13.3603 12.5628 14.2727 12.036 14.9296 12.5139L21.2889 17.1406C21.9906 17.6511 21.6818 18.7589 20.8172 18.8328L17.7404 19.0958L15.9742 21.6289C15.4779 22.3407 14.3642 22.0543 14.2729 21.1913L13.4457 13.3706ZM14.6089 13.6718L15.3388 20.5732L16.8678 18.3803C17.027 18.1519 17.2796 18.0061 17.557 17.9823L20.2207 17.7547L14.6089 13.6718Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > select field (multiple) > should have custom cursor on individual instances 1`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="select"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper go4230334221"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 8.8125C2.89645 8.8125 2.8125 8.89645 2.8125 9V10.5C2.8125 10.8107 2.56066 11.0625 2.25 11.0625C1.93934 11.0625 1.6875 10.8107 1.6875 10.5V9C1.6875 8.27513 2.27513 7.6875 3 7.6875H4.5C4.81066 7.6875 5.0625 7.93934 5.0625 8.25C5.0625 8.56066 4.81066 8.8125 4.5 8.8125H3ZM6.9375 8.25C6.9375 7.93934 7.18934 7.6875 7.5 7.6875H10.5C10.8107 7.6875 11.0625 7.93934 11.0625 8.25C11.0625 8.56066 10.8107 8.8125 10.5 8.8125H7.5C7.18934 8.8125 6.9375 8.56066 6.9375 8.25ZM12.9375 8.25C12.9375 7.93934 13.1893 7.6875 13.5 7.6875H16.5C16.8107 7.6875 17.0625 7.93934 17.0625 8.25C17.0625 8.56066 16.8107 8.8125 16.5 8.8125H13.5C13.1893 8.8125 12.9375 8.56066 12.9375 8.25ZM18.9375 8.25C18.9375 7.93934 19.1893 7.6875 19.5 7.6875H21C21.7249 7.6875 22.3125 8.27513 22.3125 9V10.6875C22.3125 10.9982 22.0607 11.25 21.75 11.25C21.4393 11.25 21.1875 10.9982 21.1875 10.6875V9C21.1875 8.89645 21.1036 8.8125 21 8.8125H19.5C19.1893 8.8125 18.9375 8.56066 18.9375 8.25ZM2.25 12.9375C2.56066 12.9375 2.8125 13.1893 2.8125 13.5V15C2.8125 15.1036 2.89645 15.1875 3 15.1875H4.125C4.43566 15.1875 4.6875 15.4393 4.6875 15.75C4.6875 16.0607 4.43566 16.3125 4.125 16.3125H3C2.27513 16.3125 1.6875 15.7249 1.6875 15V13.5C1.6875 13.1893 1.93934 12.9375 2.25 12.9375ZM21.75 13.5C22.0607 13.5 22.3125 13.7518 22.3125 14.0625V15.75C22.3125 16.0607 22.0607 16.3125 21.75 16.3125C21.4393 16.3125 21.1875 16.0607 21.1875 15.75V14.0625C21.1875 13.7518 21.4393 13.5 21.75 13.5ZM5.8125 15.75C5.8125 15.4393 6.06434 15.1875 6.375 15.1875H8.625C8.93566 15.1875 9.1875 15.4393 9.1875 15.75C9.1875 16.0607 8.93566 16.3125 8.625 16.3125H6.375C6.06434 16.3125 5.8125 16.0607 5.8125 15.75ZM10.3125 15.75C10.3125 15.4393 10.5643 15.1875 10.875 15.1875H12C12.3107 15.1875 12.5625 15.4393 12.5625 15.75C12.5625 16.0607 12.3107 16.3125 12 16.3125H10.875C10.5643 16.3125 10.3125 16.0607 10.3125 15.75Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M13.4457 13.3706C13.3603 12.5628 14.2727 12.036 14.9296 12.5139L21.2889 17.1406C21.9906 17.6511 21.6818 18.7589 20.8172 18.8328L17.7404 19.0958L15.9742 21.6289C15.4779 22.3407 14.3642 22.0543 14.2729 21.1913L13.4457 13.3706ZM14.6089 13.6718L15.3388 20.5732L16.8678 18.3803C17.027 18.1519 17.2796 18.0061 17.557 17.9823L20.2207 17.7547L14.6089 13.6718Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > select field (multiple) > should have outline 1`] = `
<div
  data-cslp="all_fields.bltapikey.en-us.select_multiple_"
>
  <p
    data-cslp="all_fields.bltapikey.en-us.select_multiple_.0"
  />
  <p
    data-cslp="all_fields.bltapikey.en-us.select_multiple_.1"
  />
</div>
`;

exports[`When an element is hovered in visual builder mode > select field (multiple) > should have outline 2`] = `
<div
  data-cslp="all_fields.bltapikey.en-us.select_multiple_"
>
  <p
    data-cslp="all_fields.bltapikey.en-us.select_multiple_.0"
  />
  <p
    data-cslp="all_fields.bltapikey.en-us.select_multiple_.1"
  />
</div>
`;

exports[`When an element is hovered in visual builder mode > select field (multiple) > should have outline 3`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go3177626044 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 34px; left: 34px; width: 828px; height: 54.3984375px;"
/>
`;

exports[`When an element is hovered in visual builder mode > select field (multiple) > should have outline on individual instances 1`] = `
<p
  data-cslp="all_fields.bltapikey.en-us.select_multiple_.0"
/>
`;

exports[`When an element is hovered in visual builder mode > select field (multiple) > should have outline on individual instances 2`] = `
<p
  data-cslp="all_fields.bltapikey.en-us.select_multiple_.0"
/>
`;

exports[`When an element is hovered in visual builder mode > select field (multiple) > should have outline on individual instances 3`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go3177626044 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 51px; left: 51px; width: 27.7734375px; height: 20.3984375px;"
/>
`;

exports[`When an element is hovered in visual builder mode > select field > should have custom cursor 1`] = `
<div
  class="go1688149084 visual-builder__cursor visible"
  data-icon="select"
  data-testid="visual-builder__cursor"
>
  <div
    class="visual-builder__cursor-wrapper go4230334221"
  >
    <div
      class="visual-builder__cursor-pointer go3046085498"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0.0167948 0.602214C-0.0784404 0.246792 0.246794 -0.0784402 0.602216 0.0167948L15.6457 4.04767C16.0914 4.1671 16.1259 4.78625 15.6962 4.95447L8.17161 7.90048C8.04753 7.94906 7.94936 8.04723 7.90078 8.17131L4.95449 15.6962C4.78627 16.1259 4.16712 16.0914 4.04769 15.6457L0.0167948 0.602214Z"
          fill="#5D50BE"
        />
      </svg>
    </div>
    <div
      class="visual-builder__cursor-icon go4261356330"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 8.8125C2.89645 8.8125 2.8125 8.89645 2.8125 9V10.5C2.8125 10.8107 2.56066 11.0625 2.25 11.0625C1.93934 11.0625 1.6875 10.8107 1.6875 10.5V9C1.6875 8.27513 2.27513 7.6875 3 7.6875H4.5C4.81066 7.6875 5.0625 7.93934 5.0625 8.25C5.0625 8.56066 4.81066 8.8125 4.5 8.8125H3ZM6.9375 8.25C6.9375 7.93934 7.18934 7.6875 7.5 7.6875H10.5C10.8107 7.6875 11.0625 7.93934 11.0625 8.25C11.0625 8.56066 10.8107 8.8125 10.5 8.8125H7.5C7.18934 8.8125 6.9375 8.56066 6.9375 8.25ZM12.9375 8.25C12.9375 7.93934 13.1893 7.6875 13.5 7.6875H16.5C16.8107 7.6875 17.0625 7.93934 17.0625 8.25C17.0625 8.56066 16.8107 8.8125 16.5 8.8125H13.5C13.1893 8.8125 12.9375 8.56066 12.9375 8.25ZM18.9375 8.25C18.9375 7.93934 19.1893 7.6875 19.5 7.6875H21C21.7249 7.6875 22.3125 8.27513 22.3125 9V10.6875C22.3125 10.9982 22.0607 11.25 21.75 11.25C21.4393 11.25 21.1875 10.9982 21.1875 10.6875V9C21.1875 8.89645 21.1036 8.8125 21 8.8125H19.5C19.1893 8.8125 18.9375 8.56066 18.9375 8.25ZM2.25 12.9375C2.56066 12.9375 2.8125 13.1893 2.8125 13.5V15C2.8125 15.1036 2.89645 15.1875 3 15.1875H4.125C4.43566 15.1875 4.6875 15.4393 4.6875 15.75C4.6875 16.0607 4.43566 16.3125 4.125 16.3125H3C2.27513 16.3125 1.6875 15.7249 1.6875 15V13.5C1.6875 13.1893 1.93934 12.9375 2.25 12.9375ZM21.75 13.5C22.0607 13.5 22.3125 13.7518 22.3125 14.0625V15.75C22.3125 16.0607 22.0607 16.3125 21.75 16.3125C21.4393 16.3125 21.1875 16.0607 21.1875 15.75V14.0625C21.1875 13.7518 21.4393 13.5 21.75 13.5ZM5.8125 15.75C5.8125 15.4393 6.06434 15.1875 6.375 15.1875H8.625C8.93566 15.1875 9.1875 15.4393 9.1875 15.75C9.1875 16.0607 8.93566 16.3125 8.625 16.3125H6.375C6.06434 16.3125 5.8125 16.0607 5.8125 15.75ZM10.3125 15.75C10.3125 15.4393 10.5643 15.1875 10.875 15.1875H12C12.3107 15.1875 12.5625 15.4393 12.5625 15.75C12.5625 16.0607 12.3107 16.3125 12 16.3125H10.875C10.5643 16.3125 10.3125 16.0607 10.3125 15.75Z"
          fill="white"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M13.4457 13.3706C13.3603 12.5628 14.2727 12.036 14.9296 12.5139L21.2889 17.1406C21.9906 17.6511 21.6818 18.7589 20.8172 18.8328L17.7404 19.0958L15.9742 21.6289C15.4779 22.3407 14.3642 22.0543 14.2729 21.1913L13.4457 13.3706ZM14.6089 13.6718L15.3388 20.5732L16.8678 18.3803C17.027 18.1519 17.2796 18.0061 17.557 17.9823L20.2207 17.7547L14.6089 13.6718Z"
          fill="white"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`When an element is hovered in visual builder mode > select field > should have outline 1`] = `
<p
  data-cslp="all_fields.bltapikey.en-us.select"
/>
`;

exports[`When an element is hovered in visual builder mode > select field > should have outline 2`] = `
<div
  class="visual-builder__hover-outline visual-builder__hover-outline--unclickable go3177626044 go1719569523"
  data-testid="visual-builder__hover-outline"
  style="top: 51px; left: 51px; width: 27.7734375px; height: 20.3984375px;"
/>
`;
